### ISSUE TITLE
Features for v3

### ISSUE THREAD

**Issue by alecthomas** [alecthomas/kingpin#90]
This issue is to track feature implementation for v3. I'm mostly happy with the core functionality in v2, but there are a few rough edges. To that end, v3 will be mostly cleanup and streamlining.

- [ ] Make hooks less ugly: `Action(*Context)` should be something like `Action(value Value)` perhaps, and called something other than Action. Not sure.
- [ ] Clean up how boolean flags work (see #54).
- [ ] Make a boolean flag that behaves more like a command than a switch? Not sure.
- [ ] Flag groups for applications with a large number of flags.
- [ ] Convert more values to use the generator.
- [ ] Some way to manipulate flags after configuration but before parsing?

**Comment by alecthomas** [comment]
Allow flags, args and command to be modified after definition. Implemented in 292d4f20095c70c7599634c58870b0567d6a6820.

**Comment by mwmahlberg** [comment]
It would be nice when kingpin used the package godoc (maybe even the function goods) to generate the manpage.

**Comment by alecthomas** [comment]
I'm not sure how that would work. What documentation would it extract? From where?

**Comment by wwalker** [comment]
+1 for "Rethink the way @ and - work"

specifically the single - by itself

**Comment by h0tbird** [comment]
+1 for comma-separated lists with autocompletion: `--roles quorum,master,worker`

**Comment by ches** [comment]
(I'm not sure if this issue is an invitation for feature requests, sorry 😇 ).

There's one thing I've wanted more than once that might fall under what you mean for for 'Help "sections"', I'm not sure: "long" help for certain items, usually subcommands. Basically, I wish for a one-line help summary that shows in `--help` (as it works today), with additional explanation that appears for `command help subcommand`. When I first started using Kingpin the `--help-long` led me to believe that may do what I want out of the box, but that isn't how it works.

In one (internal tool) case I resorted to using backtick literals for such long help, but that isn't ideal—it makes the default usage `--help` output too verbose, I manually wrapped text to sane column widths, etc. Custom template work could most likely solve this (I think?) but IMHO it seems like a use case that might be common enough to be built in. Given some time I'll try to take a stab at it, let me know if you already have some design ideas (perhaps we can open a new todo issue to keep this thread under control).

Also, +1 on `@` handling, like #65 I have also repeatedly had cases where I want to implement things like `--json @config.json`.

**Comment by alecthomas** [comment]
That's a good idea @ches.

**Comment by JoakimSoderberg** [comment]
Maybe it is already possible, but otherwise an idea would be to show boolean flags with `Default("true")`
set, as `--no-theflag` instead of `--theflag`. Since otherwise the user has to know about the fact that `--no-` is supported, or one has to write it in the description.

**Comment by alecthomas** [comment]
Yes, agreed, that would be very useful.

**Comment by Deluxo** [comment]
Also, what You think about allowing the hintAction on .Short('')? That doesn't work at the moment...

**Comment by alecthomas** [comment]
Can you clarify? Do you mean completion won't complete short flags?

**Comment by godeater** [comment]
Is it possible to have the HintAction() on one option be dependent on the value of another part of the command. e.g. I have a sub-command with a --installThisVersion flag, but the list of available versions would be dependent on something specified earlier on the same command line, like :
```
./myapp install apache_tomcat --version <tab>{"1.6", "2.4"}
./myapp install websphere --version <tab>{"6.0", "7.0"}
```

In my *actual* use case, I'd like the HintAction command to go off to a web service and query it for a list of possible completions, but I need someway to get a parameter into the HintAction function, and the value of that parameter needs to come from another part of the command line. Is this at all possible?

**Comment by ches** [comment]
@godeater This sounds like a question that would be more appropriate for someplace like Stack Overflow than this roadmap issue. I have done something similar to what you're asking with Kingpin, it's a bit ugly and I'm not entirely happy with the solution, but it does get the job done for me. I can share it in a more topical forum.

**Comment by godeater** [comment]
@ches as is ever the way, in writing up the stackoverflow question, I made this work. So I guess this isn't a feature request :)

**Comment by alecthomas** [comment]
`@` has been removed in v3-unstable.

**Comment by nthnca** [comment]
Maybe I am too late to the game to suggest features for V3. Any thoughts about trying to guess the command a user wanted? For example:

>> git shwo
git: 'shwo' is not a git command. See 'git --help'.

Did you mean this?
	show

ps. Just used kingpin for the first time. Switched from the default "flag" library. Kingpin made my application better, plus resulted in less code that was clearer to understand.

**Comment by alecthomas** [comment]
That would be quite useful, though I think it could be already possible with use of the `app.ParseContext(args)`. I'm pretty sure it has all of the information required to do this.

**Comment by jakdept** [comment]
In reference to:
>  Support for mixed mode boolean+value flags. eg. `--purge, --purge=24h`? Could use a nil value to indicate not provided? Not sure this is a good idea, plus it's hard to parse.

I'm not sure if you're looking to see if it was specifically passed bare once, or if it was passed at all. At times I see problems distinguishing between the default for a flag and that same value passed.

I personally have another flag package that adds an `.IsSet()` method to the flag. If the flag was ever set on CLI, `.IsSet()` returns true, otherwise it returns false.

I believe similar behavior could currently be achieved with `*flag.Counter() > 0`, but an `.IsSetFlag()` would likely be more obvious.

**Comment by dpedu** [comment]
Something like `LegalFileOrDirPath` would be nice. Similar to a `ExistingFileOrDir` that only accepts the path to an existing file, directory, or absent file in a directory. E.g: `/tmp/foo`, `/tmp/`, are both accepted because /tmp/ exists, but `/tmp/bar/foo` would not if `/tmp/bar` does not exist.

**Comment by elithrar** [comment]
Errors for `Required()` flags (e.g. missing value) should not only indicate the flag name, but also the environmental variable key -  e.g.

```sh
{"msg":"required flag --twitter-consumer-secret not provided","ts":"2018/01/14 21:10:52"}
```

... should be:

```
{"msg":"required flag --twitter-consumer-secret not provided (envar: "TWITTER_CONSUMER_SECRET")","ts":"2018/01/14 21:10:52"}
```

**Comment by kyoh86** [comment]
this issue is abandoned?

**Comment by davidkhala** [comment]
Will you consider release a preview or alpha version for v3 in recent future?
